File: backend/src/routes/forum.js
// src/routes/forum.js
import express from 'express';
import { authenticate } from '../middleware/auth.js';
import supabase from '../supabaseClient.js';

const router = express.Router();

/**
 * GET /api/forum/posts - Lista posts do fórum
 */
router.get('/forum/posts', authenticate, async (req, res) => {
  try {
    const { data, error } = await supabase
      .from('forum_posts')
      .select('*, profiles(full_name, avatar_url), forum_replies(count)')
      .order('created_at', { ascending: false });

    if (error) throw error;

    res.json(data);
  } catch (error) {
    console.error('Erro ao buscar posts:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/forum/posts/:id - Busca post específico com respostas
 */
router.get('/forum/posts/:id', authenticate, async (req, res) => {
  try {
    const { data: post, error: postError } = await supabase
      .from('forum_posts')
      .select('*, profiles(full_name, avatar_url)')
      .eq('id', req.params.id)
      .single();

    if (postError) throw postError;

    const { data: replies, error: repliesError } = await supabase
      .from('forum_replies')
      .select('*, profiles(full_name, avatar_url)')
      .eq('post_id', req.params.id)
      .order('created_at', { ascending: true });

    if (repliesError) throw repliesError;

    res.json({ ...post, replies });
  } catch (error) {
    console.error('Erro ao buscar post:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/forum/posts - Cria novo post
 */
router.post('/forum/posts', authenticate, async (req, res) => {
  try {
    const { data, error } = await supabase
      .from('forum_posts')
      .insert([{ ...req.body, user_id: req.user.id }])
      .select()
      .single();

    if (error) throw error;

    res.status(201).json(data);
  } catch (error) {
    console.error('Erro ao criar post:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/forum/posts/:id/replies - Adiciona resposta a um post
 */
router.post('/forum/posts/:id/replies', authenticate, async (req, res) => {
  try {
    const { data, error } = await supabase
      .from('forum_replies')
      .insert([{
        ...req.body,
        post_id: req.params.id,
        user_id: req.user.id
      }])
      .select()
      .single();

    if (error) throw error;

    res.status(201).json(data);
  } catch (error) {
    console.error('Erro ao criar resposta:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * DELETE /api/forum/posts/:id - Deleta post
 */
router.delete('/forum/posts/:id', authenticate, async (req, res) => {
  try {
    const { error } = await supabase
      .from('forum_posts')
      .delete()
      .eq('id', req.params.id)
      .eq('user_id', req.user.id);

    if (error) throw error;

    res.status(204).send();
  } catch (error) {
    console.error('Erro ao deletar post:', error);
    res.status(500).json({ error: error.message });
  }
});

export default router;

File: backend/app/routers/forum.py
from fastapi import APIRouter, Depends, HTTPException, status

from app.core.db import handle_response
from app.core.dependencies import get_current_user, get_supabase

router = APIRouter(prefix="/forum", tags=["forum"])


@router.get("/posts")
async def list_posts(user=Depends(get_current_user), supabase=Depends(get_supabase)):
    response = (
        supabase.table("forum_posts")
        .select("*, profiles(full_name, avatar_url), forum_replies(count)")
        .order("created_at", desc=True)
        .execute()
    )
    return handle_response(response)


@router.get("/posts/{post_id}")
async def retrieve_post(post_id: str, user=Depends(get_current_user), supabase=Depends(get_supabase)):
    post_response = (
        supabase.table("forum_posts")
        .select("*, profiles(full_name, avatar_url)")
        .eq("id", post_id)
        .single()
        .execute()
    )
    post = handle_response(post_response)
    if not post:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Post não encontrado")

    replies_response = (
        supabase.table("forum_replies")
        .select("*, profiles(full_name, avatar_url)")
        .eq("post_id", post_id)
        .order("created_at", desc=False)
        .execute()
    )
    replies = handle_response(replies_response) or []
    return {**post, "replies": replies}


@router.post("/posts", status_code=status.HTTP_201_CREATED)
async def create_post(payload: dict, user=Depends(get_current_user), supabase=Depends(get_supabase)):
    response = (
        supabase.table("forum_posts")
        .insert([{**payload, "user_id": user["id"]}])
        .select("*")
        .single()
        .execute()
    )
    return handle_response(response)


@router.post("/posts/{post_id}/replies", status_code=status.HTTP_201_CREATED)
async def reply_post(post_id: str, payload: dict, user=Depends(get_current_user), supabase=Depends(get_supabase)):
    response = (
        supabase.table("forum_replies")
        .insert([{**payload, "post_id": post_id, "user_id": user["id"]}])
        .select("*")
        .single()
        .execute()
    )
    return handle_response(response)


@router.delete("/posts/{post_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_post(post_id: str, user=Depends(get_current_user), supabase=Depends(get_supabase)):
    response = (
        supabase.table("forum_posts")
        .delete()
        .eq("id", post_id)
        .eq("user_id", user["id"])
        .execute()
    )
    handle_response(response)
    return {}

